catalog/video/catalog-mpegts-stream/src/main/java/com/connexta/alliance/video/stream/mpegts/PacketBuffer.java

```diff
@@ -0,0 +1,407 @@
+            currentTempFile = null;
+            bytesWrittenToTempFile = 0;
+            return Optional.of(tempFile);
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    /**
+     * Flush all buffered data to disk and rotate. Will return {@link Optional#empty()} if no data
+     * has been written to file. The caller is responsible for deleting the temp file.
+     *
+     * @return an optional temp file
+     * @throws IOException
+     */
+    public Optional<File> flushAndRotate() throws IOException {
+        lock.lock();
+        try {
+
+            if (!incompleteFrame.isEmpty()) {
+                frames.add(new Frame(FrameType.UNKNOWN, incompleteFrame));
+                incompleteFrame = new ArrayList<>();
+            }
+
+            if (!frames.isEmpty()) {
+                flushFrameset(frames.size() - 1);
+            }
+
+            return rotate(ALWAYS_TRUE);
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    private File getTempFile() throws IOException {
+        if (currentTempFile == null) {
+            tempFileCreateTime = dateSupplier.getDate()
+                    .getTime();
+            bytesWrittenToTempFile = 0;
+            currentTempFile = tempFileGenerator.generate();
+        }
+        return currentTempFile;
+    }
+
+    private Map<FrameType, AtomicLong> summarizeFrameTypes() {
+
+        Map<FrameType, AtomicLong> frameTypeCounts = new HashMap<>();
+
+        frames.forEach(frame -> {
+            frameTypeCounts.putIfAbsent(frame.frameType, new AtomicLong());
+            frameTypeCounts.get(frame.frameType)
+                    .incrementAndGet();
+        });
+
+        return frameTypeCounts;
+    }
+
+    /**
+     * Search the frame list for the last frame in a frameset. This can only be detected when the
+     * following occurs: IDR? NON-IDR* IDR. We are never guaranteed to have the leading IDR because
+     * we could start reading a stream in the middle of a frameset. If the frame list only contains
+     * UNKNOWN frame types, then always return the last index of the frame list. If the frame list
+     * contains more than {@link #maxFramesetSize}, then it is considered to be a complete framset
+     * in order to avoid memory exhaustion.
+     *
+     * @return non-null optional value, may contain index to the last frame in a frameset
+     */
+    private Optional<Integer> findLastFramesetIndex() {
+
+        if (frames.size() > maxFramesetSize) {
+            return Optional.of(frames.size() - 1);
+        }
+
+        Map<FrameType, AtomicLong> frameTypeCounts = summarizeFrameTypes();
+
+        if (frameTypeCounts.size() == 1 && frameTypeCounts.containsKey(FrameType.UNKNOWN)) {
+            return Optional.of(frames.size() - 1);
+        }
+
+        if (!frameTypeCounts.containsKey(FrameType.IDR)) {
+            return Optional.empty();
+        }
+
+        for (int i = frames.size() - 1; i >= 0; i--) {
```

Comment:
If you change the loop condition to `i > 0` you can get rid of the `if (i > 0)` in the body.